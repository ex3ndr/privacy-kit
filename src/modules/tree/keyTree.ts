import { concatBytes } from "../formats/bytes";
import { decodeUTF8, encodeUTF8 } from "../formats/text";
import { deriveKey } from "../crypto/deriveKey";
import { deriveSecureKey } from "../crypto/deriveSecureKey"
import * as crypto from 'crypto';
import * as tweetnacl from 'tweetnacl';
import { hmac_sha512 } from "../crypto/hmac_sha512";

const PACKAGE_AES_BUFFER = 0;

export class KeyTree {

    static async create(key: string | Uint8Array | Buffer, usage: string) {
        const derived = await deriveSecureKey({
            key: key,
            usage: usage
        });
        return new KeyTree(derived, usage);
    }

    #master: Uint8Array
    #usage: string

    private constructor(master: Uint8Array, usage: string) {
        this.#master = master;
        this.#usage = usage;
        Object.freeze(this);
    }

    //
    // Derive keys
    //

    deriveSymmetricKey(path: string[]): Uint8Array {
        return this.#deriveKey('aes256', [...path]);
    }

    deriveCurve25519Key(path: string[]): { secret: Uint8Array, public: Uint8Array } {
        const seed = this.#deriveKey('nacl', path);
        const keypair = tweetnacl.box.keyPair.fromSecretKey(seed);
        return {
            secret: keypair.secretKey,
            public: keypair.publicKey
        };
    }

    //
    // Hashing
    //

    hash(path: string[], data: Uint8Array|string): Uint8Array {
        const key = this.#deriveKey('hmac_sha512', [...path]);
        return hmac_sha512(key, data);
    }

    //
    // Encryption
    //

    async symmetricEncrypt(path: string[], data: Uint8Array | string): Promise<Uint8Array> {
        const key = this.deriveSymmetricKey(path);
        const nonce = crypto.randomBytes(12);
        const cipher = crypto.createCipheriv('aes-256-gcm', key, nonce);
        const encrypted = cipher.update(typeof data === 'string' ? encodeUTF8(data) : data);
        const encryptedFinal = cipher.final();
        const authTag = cipher.getAuthTag();
        return concatBytes(
            Uint8Array.from([PACKAGE_AES_BUFFER]),
            nonce,
            authTag,
            encrypted,
            encryptedFinal
        );
    }

    async symmetricDecryptBuffer(path: string[], data: Uint8Array): Promise<Uint8Array> {
        if (data[0] !== PACKAGE_AES_BUFFER) {
            throw new Error('Invalid data');
        }
        const key = this.deriveSymmetricKey(path);
        const nonce = data.subarray(1, 1 + 12);
        const authTag = data.subarray(13, 13 + 16);
        const encrypted = data.subarray(13 + 16);
        const cipher = crypto.createDecipheriv('aes-256-gcm', key, nonce);
        cipher.setAuthTag(authTag);

        const decrypted = cipher.update(encrypted);
        const decryptedFinal = cipher.final();
        return concatBytes(decrypted, decryptedFinal);
    }

    async symmetricDecryptString(path: string[], data: Uint8Array): Promise<string> {
        const decrypted = await this.symmetricDecryptBuffer(path, data);
        return decodeUTF8(decrypted);
    }

    #deriveKey(algo: string, path: string[]): Uint8Array {
        for (let p of path) {
            if (p.startsWith('#')) {
                throw new Error('Path element cannot start with #');
            }
        }
        return deriveKey(this.#master, this.#usage, [...path, '#' + algo]);
    }
}