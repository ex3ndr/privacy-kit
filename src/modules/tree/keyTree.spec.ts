import { describe, it, expect } from 'vitest';
import { KeyTree } from "./keyTree";

describe('keyTree', () => {
    it('should encrypt and decrypt', async () => {
        const keyTree = await KeyTree.create('test', 'testcase');
        const encrypted = await keyTree.symmetricEncrypt(['test'], 'test');
        const decrypted = await keyTree.symmetricDecryptString(['test'], encrypted);
        expect(decrypted).toEqual('test');
    });

    it('should detect tampering with encrypted data', async () => {
        const keyTree = await KeyTree.create('test', 'testcase');
        const encrypted = await keyTree.symmetricEncrypt(['test'], 'test');

        // Tamper with the encrypted data
        encrypted[20] = encrypted[20] ^ 1; // Flip one bit in the ciphertext

        await expect(async () => {
            await keyTree.symmetricDecryptString(['test'], encrypted);
        }).rejects.toThrow();
    });

    it('should reject invalid data format', async () => {
        const keyTree = await KeyTree.create('test', 'testcase');
        const invalidData = Buffer.from('invalid data');

        await expect(async () => {
            await keyTree.symmetricDecryptString(['test'], invalidData);
        }).rejects.toThrow('Invalid data');
    });

    it('should not decrypt with wrong key path', async () => {
        const keyTree = await KeyTree.create('test', 'testcase');
        const encrypted = await keyTree.symmetricEncrypt(['path1'], 'test');

        await expect(async () => {
            await keyTree.symmetricDecryptString(['wrong-path'], encrypted);
        }).rejects.toThrow();
    });

    it('should derive different keys for different paths', async () => {
        const keyTree = await KeyTree.create('test', 'testcase');
        const data = 'test';

        const encrypted1 = await keyTree.symmetricEncrypt(['path1'], data);
        const encrypted2 = await keyTree.symmetricEncrypt(['path2'], data);

        // Encrypted data should be different due to different keys and random nonce
        expect(encrypted1).not.toEqual(encrypted2);

        // But both should decrypt correctly with their respective paths
        const decrypted1 = await keyTree.symmetricDecryptString(['path1'], encrypted1);
        const decrypted2 = await keyTree.symmetricDecryptString(['path2'], encrypted2);

        expect(decrypted1).toEqual(data);
        expect(decrypted2).toEqual(data);
    });

    it('should produce different ciphertexts for same data and path', async () => {
        const keyTree = await KeyTree.create('test', 'testcase');
        const data = 'test';
        const path = ['test-path'];

        // Encrypt the same data multiple times
        const encrypted1 = await keyTree.symmetricEncrypt(path, data);
        const encrypted2 = await keyTree.symmetricEncrypt(path, data);
        const encrypted3 = await keyTree.symmetricEncrypt(path, data);

        // All encryptions should be different due to random nonce
        expect(encrypted1).not.toEqual(encrypted2);
        expect(encrypted2).not.toEqual(encrypted3);
        expect(encrypted1).not.toEqual(encrypted3);

        // But they should all decrypt to the same original data
        const decrypted1 = await keyTree.symmetricDecryptString(path, encrypted1);
        const decrypted2 = await keyTree.symmetricDecryptString(path, encrypted2);
        const decrypted3 = await keyTree.symmetricDecryptString(path, encrypted3);

        expect(decrypted1).toEqual(data);
        expect(decrypted2).toEqual(data);
        expect(decrypted3).toEqual(data);
    });

    it('should derive consistent symmetric keys for the same path', async () => {
        const keyTree1 = await KeyTree.create('test', 'testcase');
        const keyTree2 = await KeyTree.create('test', 'testcase');
        const path = ['test', 'path'];

        const key1 = keyTree1.deriveSymmetricKey(path);
        const key2 = keyTree2.deriveSymmetricKey(path);

        expect(key1).toEqual(key2);
        expect(key1).toMatchSnapshot();
    });

    it('should derive consistent Curve25519 keys for the same path', async () => {
        const keyTree1 = await KeyTree.create('test', 'testcase');
        const keyTree2 = await KeyTree.create('test', 'testcase');
        const path = ['test', 'path'];

        const keypair1 = keyTree1.deriveCurve25519Key(path);
        const keypair2 = keyTree2.deriveCurve25519Key(path);

        expect(keypair1.secret).toEqual(keypair2.secret);
        expect(keypair1.public).toEqual(keypair2.public);
        expect(keypair1).toMatchSnapshot();
    });

    it('should derive different keys for different usages', async () => {
        const keyTree1 = await KeyTree.create('test', 'usage1');
        const keyTree2 = await KeyTree.create('test', 'usage2');
        const path = ['test', 'path'];

        const key1 = keyTree1.deriveSymmetricKey(path);
        const key2 = keyTree2.deriveSymmetricKey(path);

        expect(key1).not.toEqual(key2);
    });
});