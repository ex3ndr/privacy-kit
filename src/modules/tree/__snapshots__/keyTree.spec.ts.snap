// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`keyTree > should derive consistent Curve25519 keys for the same path 1`] = `
{
  "public": Uint8Array [
    198,
    243,
    165,
    72,
    103,
    53,
    24,
    6,
    189,
    167,
    109,
    76,
    18,
    77,
    105,
    238,
    49,
    111,
    199,
    31,
    29,
    232,
    7,
    179,
    136,
    207,
    20,
    159,
    101,
    201,
    138,
    85,
  ],
  "secret": Uint8Array [
    188,
    91,
    162,
    9,
    21,
    2,
    65,
    174,
    92,
    124,
    27,
    130,
    26,
    22,
    208,
    253,
    226,
    227,
    44,
    232,
    187,
    159,
    233,
    196,
    41,
    156,
    225,
    43,
    179,
    53,
    174,
    61,
  ],
}
`;

exports[`keyTree > should derive consistent symmetric keys for the same path 1`] = `
Uint8Array [
  41,
  19,
  123,
  112,
  184,
  221,
  82,
  123,
  79,
  95,
  163,
  62,
  250,
  139,
  52,
  48,
  167,
  100,
  143,
  64,
  182,
  55,
  104,
  144,
  131,
  20,
  181,
  41,
  97,
  25,
  242,
  9,
]
`;
